src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,920 @@
+                    {
+                        case nameof(Enumerable.ToList):
+                            return TryConvertIfInToListInvocation(invocationExpression, out documentUpdateInfo);
+                        case nameof(Enumerable.Count):
+                            return TryConvertIfInCountInvocation(invocationExpression, out documentUpdateInfo);
+                    }
+                }
+
+                documentUpdateInfo = default;
+                return false;
+            }
+
+            private bool TryConvertIfInCountInvocation(
+                InvocationExpressionSyntax invocationExpression,
+                out DocumentUpdateInfo documentUpdateInfo)
+            {
+                if (_semanticModel.GetSymbolInfo(invocationExpression, _cancellationToken).Symbol is IMethodSymbol methodSymbol &&
+                    methodSymbol.Parameters.Length == 0 &&
+                    methodSymbol.ReturnType?.SpecialType == SpecialType.System_Int32 &&
+                    !methodSymbol.ReturnsByRef)
+                {
+                    // before var count = (from a in b select a).Count();
+                    // after
+                    // var count = 0;
+                    // foreach (var a in b)
+                    // {
+                    //     count++;
+                    //  }
+                    return TryConvertIfInInvocation(
+                            invocationExpression,
+                            (variableIdentifier, expression) => SyntaxFactory.ExpressionStatement(
+                                SyntaxFactory.PostfixUnaryExpression(SyntaxKind.PostIncrementExpression, variableIdentifier)), // Generating 'count++'
+                            SyntaxFactory.LiteralExpression(SyntaxKind.NumericLiteralExpression, SyntaxFactory.Literal(0)), // count = 0
+                            variableName: "count",
+                            out documentUpdateInfo);
+                }
+
+                documentUpdateInfo = default;
+                return false;
+            }
+
+            private bool TryConvertIfInToListInvocation(
+                InvocationExpressionSyntax invocationExpression,
+                out DocumentUpdateInfo documentUpdateInfo)
+            {
+                // before var list = (from a in b select a).ToList();
+                // after
+                // var list = new List<T>();
+                // foreach (var a in b)
+                // {
+                //     list.Add(a)
+                // }
+                if (_semanticModel.GetSymbolInfo(invocationExpression, _cancellationToken).Symbol is IMethodSymbol methodSymbol &&
+                    !methodSymbol.ReturnsByRef &&
+                    // TODO the check commented below does not work
+                    //SymbolEquivalenceComparer.Instance.Equals(methodSymbol.ReturnType.OriginalDefinition.Name, _semanticModel.Compilation.Assembly.GetTypeByMetadataName(typeof(List<>).FullName)) && 
```

Comment:
>SymbolEquivalenceComparer.Instance.Equals(methodSymbol.ReturnType.OriginalDefinition.Name, _semanticModel.Compilation.Assembly.GetTypeByMetadataName(typeof(List<>).FullName)) [](start = 22, length = 174)

it looks like this should be ```SymbolEquivalenceComparer.Instance.Equals(methodSymbol.ReturnType.OriginalDefinition, _semanticModel.Compilation.GetTypeByMetadataName(typeof(List<>).FullName))```

Reply:
Then the name check below will be redundant

---
In reply to: [178395506](https://github.com/dotnet/roslyn/pull/25362#discussion_r178395506) [](ancestors = 178395506)